core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableJoinRule.java

```diff
@@ -71,26 +71,25 @@
         EnumerableRules.LOGGER.debug(e.toString());
         return null;
       }
+    } else {
+      RelNode newRel;
+      try {
+        newRel = EnumerableHashJoin.create(
+            left,
+            right,
+            info.getEquiCondition(left, right, cluster.getRexBuilder()),
+            join.getVariablesSet(),
+            join.getJoinType());
+      } catch (InvalidRelException e) {
+        EnumerableRules.LOGGER.debug(e.toString());
+        return null;
+      }
+      if (!info.isEqui()) {
+        newRel = new EnumerableFilter(cluster, newRel.getTraitSet(),
+            newRel, info.getRemaining(cluster.getRexBuilder()));
+      }
```

Comment:
For the current implementation, the SemiJoin's join info is definitely equi, so it is okey to not shift the join keys, actually if the join only outputs oneside, we could not even add a filter here if the filter keys comes from the non-output side.

We can merge the EquiJoinInfo and NonEquiJoinInfo into JoinInfo, but this has non relationship with code snippet here. Actually for the first commit i merged them, but seems not that pointless.

I agree that we should add a check here for adding filer, although the current implementation can guarantee this.

Reply:
If I am not mistaken, the logic of this rule is:
```
if (join condition is not equi-join AND join is not INNER)
   create EnumerableThetaJoin (now EnumerableNestedLoopJoin)
else 
  create EnumerableJoin (now EnumereableHashJoin) with the equi-join part of the condition
  if (join condition is not equi-join) // i.e. it's an inner join, therefore all fields are available for the filter
     add EnumerableFilter with the non-equi-join part of the condition
```
Thus, the filter can only be added for non-equi parts of an inner join.
With this PR, we will theoretically have new scenarios:
- Non-equi SEMI and ANTI joins will go through EnumerableThetaJoin (now EnumerableNestedLoopJoin). If I am not mistaken, I think SEMI might work fine (to be verified), but ANTI will not work as expected, so probably an UnsupportedOperationException should be thrown in that case.
- Equi SEMI and ANTI joins will go through EnumerableJoin (now EnumereableHashJoin). SEMI will be supported (via BuiltInMethod.SEMI_JOIN), but ANTI will not be, so either we throw an UnsupportedOperationException or we implement it (a new BuiltInMethod.ANTI_JOIN could be simply a copy-paste of BuiltInMethod.SEMI_JOIN with one difference `return innerLookup.get().contains(key);` becomes `return !innerLookup.get().contains(key);` ). But this could be implemented as a separate PR, and for the moment the UnsupportedOperationException could be enough.

Reply:
@rubenada Thx for the analysis, i also have this concern, but Calcite does not have anti EnumerableXXX yet, without this patch, it is still not supported, do you think we have the necessity to add an UnsupportedOperationException ? I kind of curious about how Calcite support an anti join now ? Correct me if i'm wrong.

Reply:
@danny0405 If I am not mistaken, since right now there is no "JOIN type=ANTI", the antiJoin is supported via:
- either with a LEFT JOIN + a filter to retain only those elements with the right part NULL
- or with a Correlate with SemiJoinType ANTI

Reply:
@rubenada Let's add in the check when we really add in `ANTI` join expression.

Reply:
@danny0405 ok, we'll handle it afterwards